object_storage_api/services/attachment.py

```diff
@@ -91,3 +91,13 @@ def list(self, entity_id: Optional[str] = None) -> list[AttachmentMetadataSchema
         attachments = self._attachment_repository.list(entity_id)
 
         return [AttachmentMetadataSchema(**attachment.model_dump()) for attachment in attachments]
+
+    def delete(self, attachment_id: str) -> None:
+        """
+        Delete an attachment by its ID.
+        :param attachment_id: The ID of the attachment to delete.
+        """
+        stored_attachment = self._attachment_repository.get(attachment_id)
+        # Deletes attachment from object store first to prevent unreferenced objects in storage
+        self._attachment_store.delete(stored_attachment.object_key)
+        self._attachment_repository.delete(attachment_id)
```

Comment:
```suggestion
    def delete(self, attachment_id: str) -> None:
        """
        Delete an attachment by its ID.
        :param attachment_id: The ID of the attachment to delete.
        """
        stored_attachment = self._attachment_repository.get(attachment_id)
        if stored_attachment is None:
           raise MissingRecordError(detail=f"No attachment found with ID: {attachment_id}", entity_name="attachment")
        # Deletes attachment from object store first to prevent unreferenced objects in storage
        self._attachment_store.delete(stored_attachment.object_key)
        self._attachment_repository.delete(attachment_id)
```
If no changes are made to the get repo layer for attachments, then I suggest adding a `NoneType` check in the service layer to see if it exists, and raising a `MissingRecordError` (404). Otherwise the 500 error Rowan pointed out will persist, when deleting a non-existent attachment.

Reply:
@asuresh-code That is no longer the case. Please refer to my reply to Rowan [here](https://github.com/ral-facilities/object-storage-api/pull/97#issuecomment-2609535908)